### ISSUE TITLE
caproto.curio.server incompatible with curio>=1.0

### ISSUE THREAD

**Issue by pklaus** [caproto/caproto#605]
### The Problem

Recently, I was trying out the different server implementations of the caproto package. I found that the implementation using the curio package didn't work, when starting an example IOC using the following command:

```
python -m caproto.ioc_examples.simple --async-lib curio
```

It failed with an attribute error:
```
AttributeError: module 'curio' has no attribute 'KernelExit'
```

### Evaluation of the Situation

It seems that several changes have happened in curio's code base that caproto didn't keep up with:

Problem | Commit (a) | Solution
-----------|-----------|--------------
`curio.KernelExit` exception removed | dabeaz/curio@af3ff93 (Sep'18) and dabeaz/curio@60627f7 (Nov'18) | use `SystemExit`
`curio.abide()` removed | dabeaz/curio@347cc89 (Sep'19) | ? (I replaced with `curio.block_in_thread()`)
`ignore_result` kwarg removed | dabeaz/curio@53b2a42 (Apr'18) | ? (I just dropped the keyword argument)

(a): I found the relevant commits in curio's repository by searching the full text of its git log, for example by executing `git log -S KernelExit` inside curio's repository.

### Proposed Fix

1. Short Term: Pin curio package version in setup.py to `curio<1.0`
2. Long Term: Fix the curio server implementation. A first hint could be my changes to *caproto/curio/server.py* as shown below.

#### caproto diff to get caproto.curio.server up and running

```diff
diff --git a/caproto/curio/server.py b/caproto/curio/server.py
index 544df9e..a87c608 100644
--- a/caproto/curio/server.py
+++ b/caproto/curio/server.py
@@ -11,7 +11,7 @@ from ..server.common import (VirtualCircuit as _VirtualCircuit,
                              Context as _Context)
 
 
-class ServerExit(curio.KernelExit):
+class ServerExit(SystemExit):
     ...
 
 
@@ -87,7 +87,7 @@ class VirtualCircuit(_VirtualCircuit):
         # await self.pending_tasks.cancel_remaining()
 
     async def _start_write_task(self, handle_write):
-        await self.pending_tasks.spawn(handle_write, ignore_result=True)
+        await self.pending_tasks.spawn(handle_write)
 
     async def _wake_new_command(self):
         async with self.new_command_condition:
@@ -199,7 +199,7 @@ class Context(_Context):
             self._task_group = None
 
     async def _await_stop(self):
-        await curio.abide(self._stop_event.wait)
+        await curio.block_in_thread(self._stop_event.wait)
         await self._task_group.cancel_remaining()
 
     def stop(self):
```

**Comment by danielballan** [comment]
Thanks @pklaus! It looks like we do pin `curio` in our test requirements

https://github.com/caproto/caproto/blob/90e40917073dbf0da8e217622d5e5fd715257f99/test-requirements.txt#L3

but fail to pin it in our `setup.py`

https://github.com/caproto/caproto/blob/90e40917073dbf0da8e217622d5e5fd715257f99/setup.py#L38

Any interest in opening a pull request to see if your proposed changes are enough to make the tests pass on Travis-CI? If not one of us will get around to it at some point. :-)

**Comment by klauer** [comment]
I had an attempted fix in #596 but no time to update it. Either of you are welcome to take it over if you like (or start anew, of course).

**Comment by danielballan** [comment]
It turns out that re-running CI was all that was needed for #596. Merged! Will be released in caproto v0.5.1 shortly, at which point caproto will be _in_-compatible with curio < 1.2 but compatible with recent versions of curio.

Thanks for nudging this @pklaus.
